File: python/hermetix/brokers/ls.py
"""LS증권(구 이베스트투자증권) OPEN API 어댑터.

⚠️ 문서 기반 구현 (실측 전) — 공식 포털 TR 문서와 커뮤니티 카탈로그(krsec, LsApiHelper, k-ebest-im, 공식 샘플)에서
엔드포인트·TR 코드·필드명을 역추적했다. 모의서버 실측 전까지 상태는 "미검증".

규약: 모든 API 는 POST, 경로는 기능군(/stock/market-data, /stock/chart, /stock/accno, /stock/order)이고 TR 은 tr_cd 헤더로 고른다.
본문 {"<TR>InBlock": {...}} 또는 {"<TR>InBlock1": {...}}, 응답 rsp_cd("00000" 성공)/rsp_msg + OutBlock 들. 계좌번호는 토큰에 바인딩.
실전/모의 같은 호스트(모의 appkey 로 라우팅). 모의 주문은 IsuNo 에 A 접두 필수 → 항상 A+코드. HTTP 200 + rsp_cd != 00000 이 업무 오류.
TR 별 TPS(시세 3, 차트 1, 계좌 2, 예수금 1, 주문 10) → 전역 0.5s + 차트 전용 1.1s 쓰로틀.
미확인: 잔고 expcode 의 A 접두, sign 코드 의미(4·5 하락 가정), 응답 숫자 타입, 장 마감 코드(메시지 판단), medosu 표기
"""
from __future__ import annotations

import threading
import time
from datetime import datetime, timedelta, timezone
from decimal import Decimal

from ..broker import KST, BrokerClient, RateLimiter, _Http, krx_calendar, krx_tick_round
from ..errors import (
    AuthError, BrokerApiError, InsufficientFundsError, InvalidOrderError, MarketClosedError, OrderNotFoundError, RateLimitError,
)
from ..models import (
    Account, BrokerCapabilities, Candle, CandleInterval, CreateOrderRequest,
    Fill, Holding, MarketDay, Order, OrderSide, OrderStatus, OrderType, Quote, TradingEnvironment,
)
from .nh import _d, _pct

_AUTH_CODES = {"IGW00121", "IGW00123"}
_FALLING_SIGNS = {"4", "5"}


def normalize_code(raw) -> str:
    """계좌 TR 의 종목코드는 A005930 형태(추정) → 6자리 코드"""
    text = str(raw or "").strip()
    return text[1:] if len(text) == 7 and text[0] == "A" else text


class LsClient(BrokerClient):

    capabilities = BrokerCapabilities(
        broker_id="ls", market="KRX", currency="KRW",
        candle_intervals=frozenset({CandleInterval.DAY_1}),
        client_order_id=False, native_bracket=False, fractional_shares=False, server_open_orders=True,
        environments=frozenset({TradingEnvironment.PAPER, TradingEnvironment.LIVE}),
    )

    def __init__(self, app_key: str, app_secret: str, base_url: str = "https://openapi.ls-sec.co.kr:8080",
                 mac_address: str = "", exch_gubun: str = "", throttle_seconds: float = 0.5,
                 chart_throttle_seconds: float = 1.1, environment: TradingEnvironment = TradingEnvironment.PAPER):
        self.environment = environment
        self._app_key = app_key
        self._app_secret = app_secret
        self._mac_address = mac_address
        self._exch_gubun = exch_gubun
        self._http = _Http(base_url)
        self._limiter = RateLimiter(throttle_seconds, max_retries=3, backoff=lambda attempt: 1.0 * attempt)
        self._chart_limiter = RateLimiter(chart_throttle_seconds, max_retries=0)  # 차트 TR 초당 1건
        self._token: str | None = None
        self._token_expires_at = 0.0
        self._token_lock = threading.Lock()

    # ------------------------------------------------------------------ market

    def get_quotes(self, symbols: list[str]) -> list[Quote]:
        quotes = []
        for symbol in symbols:
            out = self._call("/stock/market-data", "t1102", "t1102InBlock",
                             {"shcode": self.capabilities.symbol_code(symbol), "exchgubun": self._exch_gubun}).get("t1102OutBlock") or {}
            falling = str(out.get("sign", "")) in _FALLING_SIGNS
            change, rate = _d(out.get("change")), _d(out.get("diff"))
            if falling:
                change = -change if change is not None and change > 0 else change
                rate = -rate if rate is not None and rate > 0 else rate
            quotes.append(Quote(symbol=symbol, price=_d(out.get("price")) or Decimal(0), bid_price=None, ask_price=None,
                                volume=int(_d(out.get("volume")) or 0), change=change,
                                change_rate=(rate / Decimal(100)) if rate is not None else None, timestamp=datetime.now(timezone.utc)))
        return quotes

    def get_candles(self, symbol: str, interval: CandleInterval, limit: int | None = None) -> list[Candle]:
        if interval != CandleInterval.DAY_1:
            raise ValueError("LS 어댑터는 일봉(DAY_1)만 지원합니다.")
        count = limit or 30
        today = datetime.now(KST)
        start = today - timedelta(days=count * 16 // 10 + 10)
        body = self._chart_limiter.execute(lambda: self._call("/stock/chart", "t8410", "t8410InBlock", {
            "shcode": self.capabilities.symbol_code(symbol), "gubun": "2", "qrycnt": count,
            "sdate": start.strftime("%Y%m%d"), "edate": today.strftime("%Y%m%d"), "cts_date": "", "comp_yn": "N", "sujung": "Y"}), "LS t8410")
        candles = []
        for r in body.get("t8410OutBlock1") or []:
            date = str(r.get("date", "")).strip()
            if len(date) != 8:
                continue
            candles.append(Candle(timestamp=datetime.strptime(date, "%Y%m%d").replace(tzinfo=KST),
                                  open=_d(r.get("open")) or Decimal(0), high=_d(r.get("high")) or Decimal(0),
                                  low=_d(r.get("low")) or Decimal(0), close=_d(r.get("close")) or Decimal(0),
                                  volume=int(_d(r.get("jdiff_vol")) or 0)))
        candles.sort(key=lambda c: c.timestamp)
        return candles[-count:]

    def get_calendar(self) -> list[MarketDay]:
        return krx_calendar()

    # ----------------------------------------------------------------- account

    def get_account(self) -> Account:
        s = self._balance().get("t0424OutBlock") or {}
        cash = _d(s.get("sunamt1")) or Decimal(0)  # 추정 D2 예수금
        portfolio = _d(s.get("sunamt"))
        if portfolio is None or portfolio <= 0:
            portfolio = cash + (_d(s.get("tappamt")) or Decimal(0))
        return Account(account_id=f"ls-{self.environment.value.lower()}", currency="KRW", cash=cash, portfolio_value=portfolio)

    def get_holdings(self) -> list[Holding]:
        holdings = []
        for row in self._balance().get("t0424OutBlock1") or []:
            qty = _d(row.get("janqty"))
            if qty is None or qty <= 0:
                continue
            holdings.append(Holding(symbol=normalize_code(row.get("expcode")), quantity=qty,
                                    avg_entry_price=_d(row.get("pamt")) or Decimal(0), current_price=_d(row.get("price")),
                                    market_value=_d(row.get("appamt")), unrealized_pnl=_d(row.get("dtsunik")),
                                    unrealized_pnl_rate=_pct(row.get("sunikrt"))))
        return holdings

    def get_buying_power(self) -> Decimal:
        out = self._call("/stock/accno", "CSPAQ12200", "CSPAQ12200InBlock1", {"BalCreTp": "0"}).get("CSPAQ12200OutBlock2") or {}
        return _d(out.get("MnyOrdAbleAmt")) or _d(out.get("Dps")) or Decimal(0)

    # ------------------------------------------------------------------ orders

    def create_order(self, request: CreateOrderRequest) -> Order:
        code = self.capabilities.symbol_code(request.symbol)
        is_limit = request.order_type == OrderType.LIMIT
        out = self._call("/stock/order", "CSPAT00601", "CSPAT00601InBlock1", {
            "IsuNo": f"A{code}", "OrdQty": int(request.quantity),
            "OrdPrc": int(krx_tick_round(request.limit_price)) if is_limit else 0,
            "BnsTpCode": "2" if request.side == OrderSide.BUY else "1", "OrdprcPtnCode": "00" if is_limit else "03",
            "MgntrnCode": "000", "LoanDt": "", "OrdCndiTpCode": "0"}).get("CSPAT00601OutBlock2") or {}
        order_id = str(out.get("OrdNo", "")).strip()
        if not order_id or order_id == "0":
            raise BrokerApiError(200, None, "LS 주문 응답에 OrdNo 가 없습니다")
        return Order(order_id=order_id, status=OrderStatus.SUBMITTED, symbol=code, side=request.side, order_type=request.order_type,
                     quantity=request.quantity, limit_price=request.limit_price, filled_quantity=Decimal(0),
                     client_order_id=request.client_order_id, submitted_at=datetime.now(timezone.utc))

    def get_orders(self) -> list[Order]:
        return [o for o in (self._to_order(r) for r in self._order_rows()) if o.status.is_open]

    def get_order(self, order_id: str) -> Order:
        for row in self._order_rows():
            if _same_no(row.get("ordno"), order_id):
                return self._to_order(row)
        return Order(order_id=order_id, status=OrderStatus.CANCELED)

    def cancel_order(self, order_id: str) -> Order:
        row = next((r for r in self._order_rows() if _same_no(r.get("ordno"), order_id)), None)
        if row is None:
            raise OrderNotFoundError("order-not-found", f"LS 당일 주문에서 찾을 수 없습니다: {order_id}")
        remaining = _d(row.get("ordrem"))
        if remaining is None:
            remaining = (_d(row.get("qty")) or Decimal(0)) - (_d(row.get("cheqty")) or Decimal(0))
        self._call("/stock/order", "CSPAT00801", "CSPAT00801InBlock1", {
            "OrgOrdNo": int(order_id.lstrip("0") or "0"), "IsuNo": f"A{normalize_code(row.get('expcode'))}", "OrdQty": int(remaining)})
        return Order(order_id=order_id, status=OrderStatus.PENDING_CANCEL, canceled_at=datetime.now(timezone.utc))

    def get_fills(self) -> list[Fill]:
        fills = []
        for row in self._order_rows():
            qty = _d(row.get("cheqty"))
            if qty is None or qty <= 0:
                continue
            fills.append(Fill(fill_id=None, order_id=str(row.get("ordno", "")).strip(), symbol=normalize_code(row.get("expcode")),
                              side=_side_of(row), quantity=qty, price=_d(row.get("cheprice"))))
        return fills

    # ---------------------------------------------------------------- internal

    def _balance(self) -> dict:
        return self._call("/stock/accno", "t0424", "t0424InBlock", {"prcgb": "1", "chegb": "2", "dangb": "0", "charge": "1", "cts_expcode": ""})

    def _order_rows(self) -> list[dict]:
        return self._call("/stock/accno", "t0425", "t0425InBlock",
                          {"expcode": "", "chegb": "0", "medosu": "0", "sortgb": "1", "cts_ordno": ""}).get("t0425OutBlock1") or []

    @staticmethod
    def _to_order(row: dict) -> Order:
        qty = _d(row.get("qty")) or Decimal(0)
        filled = _d(row.get("cheqty")) or Decimal(0)
        remaining = _d(row.get("ordrem"))
        if remaining is None:
            remaining = qty - filled
        text = str(row.get("status", ""))
        if "취소" in text and remaining > 0:
            status = OrderStatus.PENDING_CANCEL
        elif remaining > 0 and filled > 0:
            status = OrderStatus.PARTIALLY_FILLED
        elif remaining > 0:
            status = OrderStatus.SUBMITTED
        elif filled > 0 and filled >= qty:
            status = OrderStatus.FILLED
        elif filled > 0:
            status = OrderStatus.PARTIALLY_FILLED
        elif "거부" in text:
            status = OrderStatus.REJECTED
        else:
            status = OrderStatus.CANCELED
        price, avg = _d(row.get("price")), _d(row.get("cheprice"))
        return Order(order_id=str(row.get("ordno", "")).strip(), status=status, symbol=normalize_code(row.get("expcode")), side=_side_of(row),
                     order_type=OrderType.MARKET if str(row.get("hogagb", "")) == "03" else OrderType.LIMIT,
                     quantity=qty, limit_price=price if price and price > 0 else None,
                     filled_quantity=filled, avg_fill_price=avg if avg and avg > 0 else None)

    def _call(self, path: str, tr_cd: str, in_block: str, body: dict) -> dict:
        return self._limiter.execute(lambda: self._call_once(path, tr_cd, in_block, body), f"LS {tr_cd}")

    def _call_once(self, path: str, tr_cd: str, in_block: str, body: dict) -> dict:
        headers = {"authorization": f"Bearer {self._get_token()}", "tr_cd": tr_cd, "tr_cont": "N", "tr_cont_key": ""}
        if self._mac_address:
            headers["mac_address"] = self._mac_address
        status, parsed = self._http.request("POST", path, headers=headers, json_body={in_block: body})
        code = str(parsed.get("rsp_cd") or parsed.get("error_code") or "").strip()
        rsp_msg = str(parsed.get("rsp_msg") or parsed.get("error_description") or "")
        msg = f"LS({tr_cd}) [{code}] {rsp_msg}".strip()
        if status < 200 or status >= 300 or ("rsp_cd" in parsed and code != "00000"):
            retry_after = (getattr(self._http, "last_headers", None) or {}).get("retry-after")
            try:
                retry_after = float(retry_after) if retry_after else None
            except ValueError:
                retry_after = None
            if code == "IGW00201" or status == 429:
                raise RateLimitError(status, code, msg, retry_after)
            if code in _AUTH_CODES or status == 401:
                raise AuthError(status, code, msg)
            if any(k in rsp_msg for k in ("장종료", "장운영", "장 마감", "장마감")):
                raise MarketClosedError(status, code, msg)
            if "부족" in rsp_msg:
                raise InsufficientFundsError(status, code, msg)
            if "호가" in rsp_msg or "단위" in rsp_msg:
                raise InvalidOrderError(status, code, msg)
            if "주문번호" in rsp_msg or "원주문" in rsp_msg:
                raise OrderNotFoundError(code, msg)
            raise BrokerApiError(status, code, msg)
        return parsed

    def _get_token(self) -> str:
        if self._token and time.time() < self._token_expires_at - 600:
            return self._token
        with self._token_lock:
            if self._token and time.time() < self._token_expires_at - 600:
                return self._token
            self._limiter.throttle.wait()
            status, body = self._http.request("POST", "/oauth2/token", form_body={
                "grant_type": "client_credentials", "appkey": self._app_key, "appsecretkey": self._app_secret, "scope": "oob"})
            if status != 200 or not body.get("access_token"):
                code = body.get("error_code") or body.get("rsp_cd")
                raise AuthError(status, code, f"LS 토큰 발급 실패({code}): {body.get('error_description') or body.get('rsp_msg') or ''}")
            self._token = body["access_token"]
            self._token_expires_at = time.time() + float(body.get("expires_in", 86400))
            return self._token


def _side_of(row: dict) -> OrderSide:
    text = str(row.get("medosu", "")).strip()
    return OrderSide.BUY if "매수" in text or text == "2" else OrderSide.SELL


def _same_no(a, b) -> bool:
    return str(a or "").strip().lstrip("0") == str(b or "").strip().lstrip("0")

File: js/src/brokers/ls.ts
/**
 * LS증권(구 이베스트투자증권) OPEN API 어댑터.
 * ⚠️ 문서 기반 구현 (실측 전) — 포털 TR 문서와 커뮤니티 카탈로그에서 역추적. 모의서버 실측 전까지 "미검증".
 * - 모든 API 는 POST, 경로는 기능군(/stock/market-data, /stock/chart, /stock/accno, /stock/order)이고 TR 은 tr_cd 헤더로 고른다
 * - 본문 {"<TR>InBlock": {...}} / {"<TR>InBlock1": {...}}, 응답 rsp_cd("00000" 성공)/rsp_msg + OutBlock. 계좌번호는 토큰 바인딩
 * - 실전/모의 같은 호스트(모의 appkey 로 라우팅). 모의 주문은 IsuNo 에 A 접두 필수 → 항상 A+코드. HTTP 200 + rsp_cd != 00000 이 업무 오류
 * - TR 별 TPS(시세 3, 차트 1, 계좌 2, 예수금 1, 주문 10) → 전역 500ms + 차트 전용 1100ms 쓰로틀
 * 미확인: 잔고 expcode 의 A 접두, sign 코드(4·5 하락 가정), 응답 숫자 타입, 장 마감 코드(메시지 판단), medosu 표기
 */
import { Decimal } from "decimal.js";
import { BrokerClient, RateLimiter, httpJson, krxCalendar, krxTickRound } from "../broker.js";
import {
  AuthError, BrokerApiError, InsufficientFundsError, InvalidOrderError, MarketClosedError, OrderNotFoundError, RateLimitError,
} from "../errors.js";
import type {
  Account, BrokerCapabilities, Candle, CandleInterval, CreateOrderRequest,
  Fill, Holding, MarketDay, Order, OrderSide, OrderStatus, Quote, TradingEnvironment,
} from "../models.js";
import { symbolCodeFor } from "../models.js";

const AUTH_CODES = new Set(["IGW00121", "IGW00123"]);
const FALLING_SIGNS = new Set(["4", "5"]);

const num = (v: unknown): Decimal | null => {
  if (v === null || v === undefined) return null;
  const text = String(v).trim().replace(/,/g, "");
  if (!text) return null;
  try { return new Decimal(text); } catch { return null; }
};
const pct = (v: unknown): Decimal | null => { const r = num(v); return r ? r.div(100) : null; };
const positive = (v: unknown): Decimal | null => { const d = num(v); return d && d.gt(0) ? d : null; };

/** 계좌 TR 의 종목코드는 A005930 형태(추정) → 6자리 코드 */
export const lsNormalizeCode = (raw: unknown): string => {
  const text = String(raw ?? "").trim();
  return text.length === 7 && text[0] === "A" ? text.slice(1) : text;
};

const kstDate = (offsetDays: number): string => {
  const d = new Date(Date.now() + offsetDays * 86_400_000);
  return new Intl.DateTimeFormat("en-CA", { timeZone: "Asia/Seoul" }).format(d).replace(/-/g, "");
};
const sameNo = (a: unknown, b: unknown) => String(a ?? "").trim().replace(/^0+/, "") === String(b ?? "").trim().replace(/^0+/, "");
const sideOf = (row: Record<string, unknown>): OrderSide => {
  const t = String(row.medosu ?? "").trim();
  return t.includes("매수") || t === "2" ? "BUY" : "SELL";
};

export class LsClient implements BrokerClient {
  readonly capabilities: BrokerCapabilities = {
    brokerId: "ls", market: "KRX", currency: "KRW",
    candleIntervals: new Set<CandleInterval>(["1d"]),
    clientOrderId: false, nativeBracket: false, fractionalShares: false, serverOpenOrders: true,
    environments: new Set<TradingEnvironment>(["PAPER", "LIVE"]),
  };

  private token: string | null = null;
  private tokenExpiresAt = 0;
  private readonly limiter: RateLimiter;
  private readonly chartLimiter: RateLimiter;

  constructor(
    private readonly appKey: string,
    private readonly appSecret: string,
    readonly baseUrl: string = "https://openapi.ls-sec.co.kr:8080",
    private readonly macAddress = "",
    private readonly exchGubun = "",
    throttleMs = 500,
    chartThrottleMs = 1100,
    readonly environment: TradingEnvironment = "PAPER",
  ) {
    this.limiter = new RateLimiter(throttleMs, 3, (attempt) => 1000 * attempt);
    this.chartLimiter = new RateLimiter(chartThrottleMs, 0); // 차트 TR 초당 1건
  }

  async getQuotes(symbols: string[]): Promise<Quote[]> {
    const quotes: Quote[] = [];
    for (const symbol of symbols) {
      const out = ((await this.call("/stock/market-data", "t1102", "t1102InBlock", { shcode: symbolCodeFor(this.capabilities, symbol), exchgubun: this.exchGubun })).t1102OutBlock ?? {}) as Record<string, unknown>;
      const falling = FALLING_SIGNS.has(String(out.sign ?? ""));
      let change = num(out.change);
      let rate = num(out.diff);
      if (falling) {
        if (change && change.gt(0)) change = change.neg();
        if (rate && rate.gt(0)) rate = rate.neg();
      }
      quotes.push({ symbol, price: num(out.price) ?? new Decimal(0), bidPrice: null, askPrice: null, volume: Number(num(out.volume) ?? 0), change, changeRate: rate ? rate.div(100) : null, timestamp: new Date() });
    }
    return quotes;
  }

  async getCandles(symbol: string, interval: CandleInterval, limit?: number): Promise<Candle[]> {
    if (interval !== "1d") throw new Error("LS 어댑터는 일봉(1d)만 지원합니다.");
    const count = limit ?? 30;
    const body = await this.chartLimiter.execute(() => this.call("/stock/chart", "t8410", "t8410InBlock", {
      shcode: symbolCodeFor(this.capabilities, symbol), gubun: "2", qrycnt: count,
      sdate: kstDate(-(Math.floor(count * 16 / 10) + 10)), edate: kstDate(0), cts_date: "", comp_yn: "N", sujung: "Y",
    }), "LS t8410");
    const candles: Candle[] = [];
    for (const r of (body.t8410OutBlock1 ?? []) as Record<string, unknown>[]) {
      const date = String(r.date ?? "").trim();
      if (date.length !== 8) continue;
      candles.push({
        timestamp: new Date(`${date.slice(0, 4)}-${date.slice(4, 6)}-${date.slice(6, 8)}T00:00:00+09:00`),
        open: num(r.open) ?? new Decimal(0), high: num(r.high) ?? new Decimal(0), low: num(r.low) ?? new Decimal(0), close: num(r.close) ?? new Decimal(0),
        volume: Number(num(r.jdiff_vol) ?? 0),
      });
    }
    candles.sort((a, b) => a.timestamp.getTime() - b.timestamp.getTime());
    return candles.slice(-count);
  }

  async getCalendar(): Promise<MarketDay[]> { return krxCalendar(); }

  async getAccount(): Promise<Account> {
    const s = ((await this.balance()).t0424OutBlock ?? {}) as Record<string, unknown>;
    const cash = num(s.sunamt1) ?? new Decimal(0);
    const portfolio = positive(s.sunamt) ?? cash.plus(num(s.tappamt) ?? 0);
    return { accountId: `ls-${this.environment.toLowerCase()}`, currency: "KRW", cash, portfolioValue: portfolio, status: "ACTIVE", name: null };
  }

  async getHoldings(): Promise<Holding[]> {
    const holdings: Holding[] = [];
    for (const row of ((await this.balance()).t0424OutBlock1 ?? []) as Record<string, unknown>[]) {
      const quantity = num(row.janqty);
      if (!quantity || quantity.lte(0)) continue;
      holdings.push({ symbol: lsNormalizeCode(row.expcode), quantity, avgEntryPrice: num(row.pamt) ?? new Decimal(0), currentPrice: num(row.price), marketValue: num(row.appamt), unrealizedPnl: num(row.dtsunik), unrealizedPnlRate: pct(row.sunikrt) });
    }
    return holdings;
  }

  async getBuyingPower(): Promise<Decimal> {
    const out = ((await this.call("/stock/accno", "CSPAQ12200", "CSPAQ12200InBlock1", { BalCreTp: "0" })).CSPAQ12200OutBlock2 ?? {}) as Record<string, unknown>;
    return num(out.MnyOrdAbleAmt) ?? num(out.Dps) ?? new Decimal(0);
  }

  async createOrder(request: CreateOrderRequest): Promise<Order> {
    const code = symbolCodeFor(this.capabilities, request.symbol);
    const isLimit = request.orderType === "LIMIT";
    const out = ((await this.call("/stock/order", "CSPAT00601", "CSPAT00601InBlock1", {
      IsuNo: `A${code}`, OrdQty: request.quantity.toNumber(), OrdPrc: isLimit ? krxTickRound(request.limitPrice!).toNumber() : 0,
      BnsTpCode: request.side === "BUY" ? "2" : "1", OrdprcPtnCode: isLimit ? "00" : "03", MgntrnCode: "000", LoanDt: "", OrdCndiTpCode: "0",
    })).CSPAT00601OutBlock2 ?? {}) as Record<string, unknown>;
    const orderId = String(out.OrdNo ?? "").trim();
    if (!orderId || orderId === "0") throw new BrokerApiError(200, null, "LS 주문 응답에 OrdNo 가 없습니다");
    return { orderId, status: "SUBMITTED", symbol: code, side: request.side, orderType: request.orderType, quantity: request.quantity, limitPrice: request.limitPrice ?? null, filledQuantity: new Decimal(0), clientOrderId: request.clientOrderId ?? null, submittedAt: new Date() };
  }

  async getOrders(): Promise<Order[]> {
    return (await this.orderRows()).map((r) => this.toOrder(r)).filter((o) => o.status === "SUBMITTED" || o.status === "PARTIALLY_FILLED" || o.status === "PENDING_CANCEL");
  }

  async getOrder(orderId: string): Promise<Order> {
    const row = (await this.orderRows()).find((r) => sameNo(r.ordno, orderId));
    return row ? this.toOrder(row) : { orderId, status: "CANCELED" };
  }

  async cancelOrder(orderId: string): Promise<Order> {
    const row = (await this.orderRows()).find((r) => sameNo(r.ordno, orderId));
    if (!row) throw new OrderNotFoundError("order-not-found", `LS 당일 주문에서 찾을 수 없습니다: ${orderId}`);
    const remaining = num(row.ordrem) ?? (num(row.qty) ?? new Decimal(0)).minus(num(row.cheqty) ?? 0);
    await this.call("/stock/order", "CSPAT00801", "CSPAT00801InBlock1", { OrgOrdNo: Number(orderId.replace(/^0+/, "") || "0"), IsuNo: `A${lsNormalizeCode(row.expcode)}`, OrdQty: remaining.toNumber() });
    return { orderId, status: "PENDING_CANCEL", canceledAt: new Date() };
  }

  async getFills(): Promise<Fill[]> {
    return (await this.orderRows()).filter((r) => (num(r.cheqty) ?? new Decimal(0)).gt(0))
      .map((r) => ({ fillId: null, orderId: String(r.ordno ?? "").trim(), symbol: lsNormalizeCode(r.expcode), side: sideOf(r), quantity: num(r.cheqty), price: num(r.cheprice) }));
  }

  private balance(): Promise<Record<string, unknown>> {
    return this.call("/stock/accno", "t0424", "t0424InBlock", { prcgb: "1", chegb: "2", dangb: "0", charge: "1", cts_expcode: "" });
  }

  private async orderRows(): Promise<Record<string, unknown>[]> {
    return ((await this.call("/stock/accno", "t0425", "t0425InBlock", { expcode: "", chegb: "0", medosu: "0", sortgb: "1", cts_ordno: "" })).t0425OutBlock1 ?? []) as Record<string, unknown>[];
  }

  private toOrder(row: Record<string, unknown>): Order {
    const qty = num(row.qty) ?? new Decimal(0);
    const filled = num(row.cheqty) ?? new Decimal(0);
    const remaining = num(row.ordrem) ?? qty.minus(filled);
    const text = String(row.status ?? "");
    let status: OrderStatus;
    if (text.includes("취소") && remaining.gt(0)) status = "PENDING_CANCEL";
    else if (remaining.gt(0) && filled.gt(0)) status = "PARTIALLY_FILLED";
    else if (remaining.gt(0)) status = "SUBMITTED";
    else if (filled.gt(0) && filled.gte(qty)) status = "FILLED";
    else if (filled.gt(0)) status = "PARTIALLY_FILLED";
    else if (text.includes("거부")) status = "REJECTED";
    else status = "CANCELED";
    return { orderId: String(row.ordno ?? "").trim(), status, symbol: lsNormalizeCode(row.expcode), side: sideOf(row), orderType: String(row.hogagb ?? "") === "03" ? "MARKET" : "LIMIT", quantity: qty, limitPrice: positive(row.price), filledQuantity: filled, avgFillPrice: positive(row.cheprice) };
  }

  private call(path: string, trCd: string, inBlock: string, input: Record<string, unknown>): Promise<Record<string, unknown>> {
    return this.limiter.execute(() => this.callOnce(path, trCd, inBlock, input), `LS ${trCd}`);
  }

  private async callOnce(path: string, trCd: string, inBlock: string, input: Record<string, unknown>): Promise<Record<string, unknown>> {
    const headers: Record<string, string> = { "Content-Type": "application/json; charset=utf-8", authorization: `Bearer ${await this.getToken()}`, tr_cd: trCd, tr_cont: "N", tr_cont_key: "" };
    if (this.macAddress) headers.mac_address = this.macAddress;
    const [status, parsed, resHeaders] = await httpJson(this.baseUrl + path, { method: "POST", headers, body: JSON.stringify({ [inBlock]: input }) });
    const code = String(parsed.rsp_cd ?? parsed.error_code ?? "").trim();
    const rspMsg = String(parsed.rsp_msg ?? parsed.error_description ?? "");
    const msg = `LS(${trCd}) [${code}] ${rspMsg}`.trim();
    if (status < 200 || status >= 300 || ("rsp_cd" in parsed && code !== "00000")) {
      const retryAfter = Number(resHeaders?.["retry-after"]);
      if (code === "IGW00201" || status === 429) throw new RateLimitError(status, code, msg, Number.isFinite(retryAfter) ? retryAfter : null);
      if (AUTH_CODES.has(code) || status === 401) throw new AuthError(status, code, msg);
      if (["장종료", "장운영", "장 마감", "장마감"].some((k) => rspMsg.includes(k))) throw new MarketClosedError(status, code, msg);
      if (rspMsg.includes("부족")) throw new InsufficientFundsError(status, code, msg);
      if (rspMsg.includes("호가") || rspMsg.includes("단위")) throw new InvalidOrderError(status, code, msg);
      if (rspMsg.includes("주문번호") || rspMsg.includes("원주문")) throw new OrderNotFoundError(code, msg);
      throw new BrokerApiError(status, code, msg);
    }
    return parsed;
  }

  private async getToken(): Promise<string> {
    if (this.token && Date.now() < this.tokenExpiresAt - 600_000) return this.token;
    await this.limiter.throttle.wait();
    const form = new URLSearchParams({ grant_type: "client_credentials", appkey: this.appKey, appsecretkey: this.appSecret, scope: "oob" });
    const [status, body] = await httpJson(`${this.baseUrl}/oauth2/token`, { method: "POST", headers: { "Content-Type": "application/x-www-form-urlencoded" }, body: form.toString() });
    if (status !== 200 || typeof body.access_token !== "string") {
      const code = String(body.error_code ?? body.rsp_cd ?? "");
      throw new AuthError(status, code || null, `LS 토큰 발급 실패(${code}): ${body.error_description ?? body.rsp_msg ?? ""}`);
    }
    this.token = body.access_token;
    this.tokenExpiresAt = Date.now() + Number(body.expires_in ?? 86400) * 1000;
    return this.token;
  }
}
